### ISSUE TITLE
regression: uv build silently produces empty wheels with hatchling if .gitignore is missing

### ISSUE THREAD

**Issue by jinnatar** [astral-sh/uv#8200]
### Summary:
Newer versions of uv build "empty" wheels under the following conditions with `hatchling.build` when depending on hatch default logic of finding the sources at `src/$project/__init__.py`.
- Last version where it works: `0.4.17`
- First version that fails: `0.4.18`
- The trigger seems to be whether the project root has a file `.gitignore` or not, where it being missing triggers this issue!

Minimal pyproject.toml:
```toml
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "demo"
version = "0.1.1"
description = ""
requires-python = ">=3.11"
dependencies = []

[project.scripts]
demo = 'demo:run'
```

Source contents is only one file at `src/demo/__init__.py`:
```python
def run():
    print("Running like the wind!")
```

Dockerfile for repro:
```docker
# Fails with 0.4.18
FROM ghcr.io/astral-sh/uv:0.4.18-python3.12-alpine AS builder
# Succeeds with 0.4.17
#FROM ghcr.io/astral-sh/uv:0.4.17-python3.12-alpine AS builder

RUN mkdir /build
WORKDIR /build

COPY pyproject.toml uv.lock ./
COPY src ./src
RUN uv build

CMD ["unzip", "-l", "dist/demo-0.1.1-py3-none-any.whl"]
```

Expected output:
```
Archive:  dist/demo-0.1.1-py3-none-any.whl
  Length      Date    Time    Name
---------  ---------- -----   ----
       47  02-02-2020 00:00   demo/__init__.py
       72  02-02-2020 00:00   demo-0.1.1.dist-info/METADATA
       87  02-02-2020 00:00   demo-0.1.1.dist-info/WHEEL
       34  02-02-2020 00:00   demo-0.1.1.dist-info/entry_points.txt
      358  02-02-2020 00:00   demo-0.1.1.dist-info/RECORD 
 --------                     -------                  
      598                     5 files
```

With newer versions instead we get:
```
Archive:  dist/demo-0.1.1-py3-none-any.whl
  Length      Date    Time    Name
---------  ---------- -----   ----
       72  02-02-2020 00:00   demo-0.1.1.dist-info/METADATA
       87  02-02-2020 00:00   demo-0.1.1.dist-info/WHEEL
       34  02-02-2020 00:00   demo-0.1.1.dist-info/entry_points.txt
      287  02-02-2020 00:00   demo-0.1.1.dist-info/RECORD
 --------                     -------
      480                     4 files
```

- The key diff is that the `demo` package itself is missing, we only get the dist-info, which while technically a valid wheel, doesn't do anything useful either.

**Comment by jinnatar** [comment]
What's confusing here is that in my local dev env, `uv build` with `0.4.21` produces a valid wheel with src contents. So not sure what bit in the Docker env minimalism makes it trip.

**Comment by jinnatar** [comment]
No one will believe me, BUT, the presence of a `.gitignore` file in the build context, even if empty, makes it work. If it's not there, as it's not in my minimal Dockerfile, it fails.

**Comment by jinnatar** [comment]
No useful verbose logging diff between the failing and working builds:
```diff
32d31
< DEBUG Found fresh response for: https://files.pythonhosted.org/packages/88/5f/e351af9a41f866ac3f1fac4ca0613908d9a41741cfcf2228f4ad853b697d/pluggy-1.5.0-py3-none-any.whl.metadata
33a33
> DEBUG Found fresh response for: https://files.pythonhosted.org/packages/88/5f/e351af9a41f866ac3f1fac4ca0613908d9a41741cfcf2228f4ad853b697d/pluggy-1.5.0-py3-none-any.whl.metadata
41c41
< DEBUG Installing in hatchling==1.25.0, packaging==24.1, pathspec==0.12.1, pluggy==1.5.0, trove-classifiers==2024.10.13 in /home/user/.cache/uv/builds-v0/.tmpnuqDDO
---
> DEBUG Installing in hatchling==1.25.0, packaging==24.1, pathspec==0.12.1, pluggy==1.5.0, trove-classifiers==2024.10.13 in /home/user/.cache/uv/builds-v0/.tmpE02Jyf
50c50
< DEBUG Calling `hatchling.build.build_sdist("/vol/home/src/uv-repro.git/dist/.tmpigoeaQ", {})`
---
> DEBUG Calling `hatchling.build.build_sdist("/vol/home/src/uv-repro.git/dist/.tmpmpoSRe", {})`
72,73c72,73
< DEBUG Split specific environment resolution took 0.000s
< DEBUG Installing in hatchling==1.25.0, packaging==24.1, pathspec==0.12.1, pluggy==1.5.0, trove-classifiers==2024.10.13 in /home/user/.cache/uv/builds-v0/.tmpsS489a
---
> DEBUG Split specific environment resolution took 0.001s
> DEBUG Installing in hatchling==1.25.0, packaging==24.1, pathspec==0.12.1, pluggy==1.5.0, trove-classifiers==2024.10.13 in /home/user/.cache/uv/builds-v0/.tmptTLjxm
82c82
< DEBUG Calling `hatchling.build.build_wheel("/vol/home/src/uv-repro.git/dist/.tmpUJojVo", {}, None)`
---
> DEBUG Calling `hatchling.build.build_wheel("/vol/home/src/uv-repro.git/dist/.tmp3We1rG", {}, None)`
```

**Comment by jinnatar** [comment]
My best guess is, it's something to do with https://github.com/astral-sh/uv/pull/7835 @konstin does this make any sense to you as the source of the regression?

**Comment by charliermarsh** [comment]
Thanks @jinnatar -- we're taking a look!

**Comment by charliermarsh** [comment]
(@konstin has been debugging to try and understand why hatchling is behaving this way.)

**Comment by zanieb** [comment]
Related https://github.com/pypa/hatch/issues/1273

**Comment by konstin** [comment]
I've added a fix and an explanation in #8220, it's a chain of circumstances adding together.

Thanks for the great write-up and reproducer, they were very helpful for debugging.
